Adding System.Tuples

jcouv opened the pull request:
Added the ValueTuple types that we are discussing for C# 7.

The code is based off [System.Tuple](http://referencesource.microsoft.com/#mscorlib/system/tuple.cs,9124c4bea9ab0199) and the tests are based off [TupleTests](https://github.com/dotnet/corefx/blob/cd5d2ea3080927b09022adf06745c1876217c3ca/src/System.Runtime/tests/System/Tuple.cs).

A few open issues left:
- The hash function is rather poor and needs further discussion.
- I removed `[Serializable]` as it doesn't appear anywhere in corefx.
- the solution and project are named after the package that we discussed `System.Tuples`, but should be probably renamed to `System.ValueTuples` when moved to shipping project.
- Pattern matching may not need the `ITuple` interface after all, so I did no include it.
- There is a small difference between `Tuple.ToString` and `ValueTuple.ToString`, where one prints `(1, ..., 7, (8, 9))`, but the latter prints `(1, ..., 7, 8, 9)`.
- I included xml docs only on ValueTuple\`2. 

@weshaggard for review.

CC @dotnet/roslyn-compiler @tmat as FYI.

tmat commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        bool IStructuralEquatable.Equals(object other, IEqualityComparer comparer)
+        {
+            if (other == null || !(other is ValueTuple)) return false;
```
```return other is ValueType;```

jcouv replied:
Done.

tmat commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        string ITupleInternal.ToString(StringBuilder sb)
+        {
+            return "()";
```
```return ToString();``` to avoid duplication of literal

jcouv replied:
There is actually a bug here. This method should append to `sb`. I'll add a test and fix.

tmat commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+            return "()";
+        }
+
+        int ITupleInternal.Size
```
```=> 0;```

jcouv replied:
Done.

tmat commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+            return sb.ToString();
+        }
+
+        int ITupleInternal.Size
```
```=> 1;```

SunnyWar replied:
Why multiple calls to Append?  Matter of fact, why _any _calls to Append?

return sb.ToString() + $"{Item1})";

jcouv replied:
@tmat. Done. I replaced all the constant getters with arrow syntax.

jcouv replied:
@SunnyWar The reason for using `StringBuilder` and `Append` is to economize on short-lived object allocations.

tmat commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+            int c = 0;
+
+            c = comparer.Compare(Item1, objTuple.Item1);
```
```int c = comparer.Compare(Item1, objTuple.Item1);```

jcouv replied:
Done.

tmat commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+    /// </summary>
+    internal interface ITupleInternal
+    {
+        string ToString(StringBuilder sb);
```
Feels like this should just be ```void AppendTo(StringBuilder sb)``` and the caller should call ToString() on the result.

jcouv replied:
Done.

mellinoe commented:
How are we thinking about this in regards to the .NET Framework? Are we going to ship this in-box there, or will you need to reference this NuGet package if you want to use the language feature? It seems like we should prefer the former, as I'm not aware of any other language features that require an outside component to function. If that's the case, we will need to complicate things a bit further by adding a "ref" project, and a facade configuration for the in-box assembly.

stephentoub commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        public override bool Equals(object obj)
+        {
+            return ((IStructuralEquatable)this).Equals(obj, EqualityComparer<object>.Default);
```
Is there special runtime support for this, or does this end up boxing ```this```?

jcouv replied:
Jared confirmed that the cast to interface will box. I'll change. I think it can be simplified to:
```
            return obj is ValueTuple<T1> && Equals((ValueTuple<T1>)obj);
```

stephentoub commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+            var objTuple = (ValueTuple<T1>)other;
+
+            return comparer.Equals(Item1, objTuple.Item1);
```
Similarly, is there anything better that can be done here?  Seems like, for example, calling ```Equals(obj)``` on a ```ValueTuple<int>```, for example, will not only box the ```ValueTuple<int>```, but will also box the ```int``` here.

jcouv replied:
I feel stuck on this one. Both `Item1` and `objTuple.Item1` get boxed because `comparer.Equals` takes two objects.
We do have better comparers elsewhere (`IEqualityComparer<T>`) but here I'm restricted by what the `IStructuralEquatable` interface gives me...

stephentoub commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+        {
+            sb.Append(Item1);
+            sb.Append(")");
+            return sb.ToString();
```
I don't understand why this approach with StringBuilder is being used.  Is this all because of the TRest option on the ```ValueTuple<T1, T2, T3, T4, T5, T6, T7, TRest>```?  Seems like we should really be exploring an alternative, as I imagine that's a relatively rare case, and yet we're paying non-trivially for it, e.g. calling ```ValueTuple<T1, T2>.ToString()``` should be as simple as ```return "(" + Item1 + ", " + Item2 + ")";```, which will use string.Join and incur the overhead of the implicit ```string[]``` allocation and the resulting string, whereas using StringBuilder like this will allocate the StringBuilder, the StringBuilder's internal ```char[]``` (if not more than one if it needs to grow), etc.

jcouv replied:
Yes, short tuples should be the most common case. I'll change.
You may want to take a look at the existing `System.Tuples` to see if you want to make a similar change.

stephentoub commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+        /// <remarks>
+        /// The <paramref name="other"/> parameter is considered to be equal to the current instance under the following conditions:
+        /// <list type="bullet">
+        ///     <item><description>It is a <seealso cref="Tuple&lt;T1, T2&gt;"/> struct.</description></item>
```
Can you use the ```Tuple{T1, T2}``` syntax instead of using ```Tuple&lt;T1, T2&gt;```?  And shouldn't all of these be ```ValueTuple```s in the comments rather than ```Tuple```?

jcouv replied:
I didn't know about the curly braces syntax. I'll change that and fix the mistaken references to `Tuple`.

jcouv replied:
Done.

JonHanna commented on src/System.Tuples/tests/Tuples/UnitTests.cs:
```diff
@@ -0,0 +1,797 @@
+            return x.GetHashCode();
+        }
+    }
+}
```
Nit: Newline at end of file.

jcouv replied:
Done

jcouv commented:
@mellinoe 
Based on discussion with Immo, Krysztof and Wes, the long-term shipping vehicles we determined are mscorlib (CLR and CoreCLR) and System.Private.Corlib (CoreRT). 
The current assembly and package (System.Tuples) are for experimentation purpose.

I forwarded you the thread.

stephentoub commented:
> Based on discussion with Immo, Krysztof and Wes, the long-term shipping vehicles we determined are mscorlib (CLR and CoreCLR) and System.Private.Corlib (CoreRT). The current assembly and package (System.Tuples) are for experimentation purpose.

In that case, is this the right repo to put this in?  mscorlib is in the coreclr repo, and System.Private.Corlib is in the corert repo.  And if this is experimental, seems like it should start in the corefxlab repo?

stephentoub commented:
Or, are we expecting it'll be exposed from the System.Runtime contract, for example, have the code live in corefx, and make System.Runtime a partial facade for the .NET Framework and corert?

jcouv commented:
@mellinoe and @stephentoub, regarding the choice of repo, you may want to discuss with Wes. He concluded that is the right place after our meeting (forwarded you the notes). I'm happy to move if needed.

mellinoe commented:
Unless I'm missing something, we should be able to implement this directly in System.Runtime (the partial facade) for both CoreCLR and CoreRT, rather than forward this to mscorlib or System.Private.CoreLib for either of those platforms. We shouldn't need to implement this in mscorlib unless other types in mscorlib need to rely on it, but I don't think that's the case. Ditto for System.Private.CoreLib.

For the .NET Framework, we are pretty much stuck with implementing it in mscorlib unless we want to require you to download a NuGet package to use the language feature.

mellinoe commented:
@jcouv Thanks for the extra context. Unless I'm mistaken about the above ^, it seems like we can simplify things w.r.t. implementation assembly, and eventually put this into System.Runtime.dll directly (which does live in this repo).

jcouv commented:
@mellinoe and @stephentoub, I just remembered an additional factor which I think motivated the choice of corlib: we plan to have an `ITuple` interface (with simple size and indexer) which other existing types should implement (`System.Tuple`, `KeyValuePair` and other tuple-like types).

I didn't include `ITuple` yet, because the language design is changing. I will check with the folks working on pattern matching to solidify the plan.

weshaggard commented:
The goal will likely be to put at least part of this into System.Runtime long term but while this is in an experimental state we want to keep it as a standalone library while it churns. Keeping it in its own library allows us to play with this new feature on .NET Core and on the existing .NET Framework. Once we believe the design is final and the language feature is implemented then we will likely move the implementations into mscorlib/corelib so that we can implement the new ITuple interface on the existing Tuple type, but even when we do that we might still keep ValueTuple in this standalone library and implement it as a partial facade for the newer frameworks and likely still standalone for the older framework support.

weshaggard commented on src/System.Tuples/pkg/System.Tuples.builds:
```diff
@@ -0,0 +1,8 @@
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <ItemGroup>
+    <Project Include="System.Tuples.pkgproj"/>
```
Please do rename this to System.ValueTuple before commiting.

jcouv replied:
Done

weshaggard replied:
I see you renamed the library, please also rename the directory to match.

jcouv replied:
Done. Sorry about that.

weshaggard commented on src/System.Tuples/src/System.Tuples.csproj:
```diff
@@ -0,0 +1,25 @@
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <ProjectGuid>{4C2655DB-BD9E-4C86-83A6-744ECDDBDF29}</ProjectGuid>
+    <PackageTargetFramework>netstandard1.1</PackageTargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Do you actually use unsafe code here? If not please remove this.

jcouv replied:
Done.

weshaggard commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        public bool Equals(ValueTuple other)
+        {
+            return true;
```
Why is this always true?

weshaggard replied:
Never mind I see this is the empty ValueTuple.

jaredpar commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        public override int GetHashCode()
+        {
+            return ((IStructuralEquatable)this).GetHashCode(EqualityComparer<object>.Default);
```
This causes a box to occur.  Can we call the method directly?

jaredpar commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        int ITupleInternal.GetHashCode(IEqualityComparer comparer)
+        {
+            return ((IStructuralEquatable)this).GetHashCode(comparer);
```
This causes a box to occur.  Can we call the method directly?

jaredpar commented on src/System.Tuples/src/System/Tuples/ValueTuple.cs:
```diff
@@ -0,0 +1,1205 @@
+
+        int IComparable.CompareTo(object obj)
+        {
+            return ((IStructuralComparable)this).CompareTo(obj, Comparer<object>.Default);
```
Hmm another box.  We should revisit this pattern through the entire code base.  Otherwise simple struct operations are going to end up doing a lot of allocations.

jcouv replied:
Agreed. Let me see how to do that.

jcouv commented:
@mellinoe @stephentoub @tmat @weshaggard @jaredpar 
Pushed 3 commits that address the feedback so far.

I was not able to reduce boxing for methods that implement `IStructuralComparable` or `IStructuralEquatable`. But `object.Equals`, `IComparable.CompareTo` and `object.GetHashCode` should be better now on `ValueTuple'1` and `ValueTuple'8` (if those are good I will apply the same pattern across).

Would any of you know `GetHashCode` on existing `System.Tuple'8` doesn't hash all elements? I'm thinking to change that here as hashing all elements would simplify the code.

Also, I'm thinking of a way to reduce the number of static fields I just added. Maybe only keep one of each.

stephentoub commented on src/System.Tuples/src/project.json:
```diff
@@ -0,0 +1,18 @@
+﻿{
+  "dependencies": {
+    "Microsoft.NETCore.Platforms": "1.0.1-rc3-24011-00",
```
FYI, the CI is failing due to verification that all of the project.jsons use the desired version.  The version gets updated daily, and this will need to be kept in sync.

stephentoub commented on src/System.Tuples/tests/project.json:
```diff
@@ -0,0 +1,20 @@
+    "System.Runtime": "4.1.0-rc3-24011-00",
+    "System.Runtime.Extensions": "4.1.0-rc3-24011-00",
+    "System.Text.RegularExpressions": "4.0.12-rc3-24011-00",
+    "System.Threading": "4.0.11-rc3-24011-00",
```
Same comment about updating these versions.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+    {
+        public override bool Equals(object obj)
+        {
+            return ((IStructuralEquatable)this).Equals(obj, EqualityComparer<object>.Default);
```
Can this instead be:
```C#
return obj is ValueTuple;
```
?

jcouv replied:
Thanks. Done.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+
+        int IComparable.CompareTo(object obj)
+        {
+            return ((IStructuralComparable)this).CompareTo(obj, Comparer<object>.Default);
```
Rather than delegating to IStructuralComparable.Compare, which ends up boxing ```this```, you could either copy the implementation from that method to here, or you could put that implementation into a private helper that's called by both interface implementations.

jcouv replied:
Done

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+
+        int IStructuralEquatable.GetHashCode(IEqualityComparer comparer)
+        {
+            return comparer.GetHashCode();
```
Why are we using the comparer's hash code?  Should this just be ```return 0;``` or something?

jcouv replied:
Good catch. Thanks. Fixed.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+
+        public override int GetHashCode()
+        {
+            return ((IStructuralEquatable)this).GetHashCode(EqualityComparer<object>.Default);
```
There's no need to box here.  It could instead be:
```C#
return EqualityComparer<object>.Default.GetHashCode();
```
or if you decide to change the hash code based on my comment below, just:
```C#
return 0;
```

jcouv replied:
Fixed to 0.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        public T1 Item1;
+
+        private static readonly EqualityComparer<T1> s_equality1 = EqualityComparer<T1>.Default;
+        private static readonly Comparer<T1> s_comparer1 = Comparer<T1>.Default;
```
Why have these static fields?  The comparers are already cached in the Default implementations.

jcouv replied:
I went back and forth on this. I'm happy to inline call to Default.

VSadov replied:
In similar ValueTuple that we use in Roslyn we cache the comparers. 
We have found that calling static generic properties like Default is pretty expensive while static field accesses are inlined by the JIT so essentially noops.

I think we should cache the comparers, but that can also be done later if profiling indicates the need (I think it will).

stephentoub replied:
If there's really an issue here with the JIT, I'd rather see us address that in the JIT and/or in the implementation of these Default properties rather than proliferate such static fields.
cc: @jkotas, @CarolEidt, @cmckinsey

KrzysztofCwalina replied:
But, I think we should use fields for now, file an issue for the JIT, add TODOs to remove the fields when JIT issues is addressed.

stephentoub replied:
I'm not convinced.  How often are ValueTuple's going to be compared for equality?  I have to believe that any difference here would only be evident in a really tight microbenchmark, and would that matter for anything in the interim?  In contrast, we'll end up paying for these fields in memory even when equality isn't used.

KrzysztofCwalina replied:
I guess it depends how soon we can fix the JIT. I do believe that some developers will compare these frequently and on the hot path.

jkotas replied:
Doing something about this in the JIT is hard. Hops between generic types are always going to have some cost because of generic code sharing.

stephentoub replied:
@jkotas, thanks.  The original implementation here had a field on ```ValueTuple<T1>``` that was then used by each of the Equals methods on the various types, e.g.
```C#
return 
    ValueTuple<T1>.s_comparer.Equals(...) &&
    ValueTuple<T2>.s_comparer.Equals(...) &&
    ...
```
Is that significantly different from a JIT perspective than using EqualityComparer<T1>.Default instead?

I tried the following microbenchmark (which I of course could have messed up, so please let me know if I did).  It has three implementations: a ```ValueTuple_SharedField<T1, T2>``` that uses a single cached ```ValueTuple<T>.s_comparer``` field, a ```ValueTuple_LocalField<T1, T2>``` that has its own set of cached fields for each of T1 and T2, and a ```ValueTuple_Default<T1, T2>``` that uses ```EqualityComparer<T>.Default```.
```C#
using System;
using System.Collections.Generic;
using System.Diagnostics;

class Program
{
    static void Main()
    {
        var sw = new Stopwatch();
        var vtf1 = new ValueTuple_SharedField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtf2 = new ValueTuple_SharedField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtl1 = new ValueTuple_LocalField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtl2 = new ValueTuple_LocalField<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtd1 = new ValueTuple_Default<int, double> { Item1 = 42, Item2 = 42.0 };
        var vtd2 = new ValueTuple_Default<int, double> { Item1 = 42, Item2 = 42.0 };

        const int Iters = 100000000;
        bool result = true;
        while (true)
        {
            sw.Restart();
            for (int i = 0; i < Iters; i++) result |= vtf1.Equals(vtf2);
            sw.Stop();
            Console.WriteLine($"Shared:  {sw.Elapsed.TotalSeconds}");

            sw.Restart();
            for (int i = 0; i < Iters; i++) result |= vtl1.Equals(vtl2);
            sw.Stop();
            Console.WriteLine($"Local:   {sw.Elapsed.TotalSeconds}");

            sw.Restart();
            for (int i = 0; i < Iters; i++) result |= vtd1.Equals(vtd1);
            sw.Stop();
            Console.WriteLine($"Default: {sw.Elapsed.TotalSeconds}");

            Console.WriteLine();
        }
    }
}

struct ValueTuple<T1> { internal static EqualityComparer<T1> s_comparer = EqualityComparer<T1>.Default; }

struct ValueTuple_SharedField<T1, T2> : IEquatable<ValueTuple_SharedField<T1, T2>>
{
    public T1 Item1;
    public T2 Item2;

    public bool Equals(ValueTuple_SharedField<T1, T2> other) =>
        ValueTuple<T1>.s_comparer.Equals(Item1, other.Item1) &&
        ValueTuple<T2>.s_comparer.Equals(Item2, other.Item2);
}

struct ValueTuple_LocalField<T1, T2> : IEquatable<ValueTuple_LocalField<T1, T2>>
{
    internal static readonly EqualityComparer<T1> s_comparer1 = EqualityComparer<T1>.Default;
    internal static readonly EqualityComparer<T2> s_comparer2 = EqualityComparer<T2>.Default;

    public T1 Item1;
    public T2 Item2;

    public bool Equals(ValueTuple_LocalField<T1, T2> other) =>
        s_comparer1.Equals(Item1, other.Item1) &&
        s_comparer2.Equals(Item2, other.Item2);
}

struct ValueTuple_Default<T1, T2> : IEquatable<ValueTuple_Default<T1, T2>>
{
    public T1 Item1;
    public T2 Item2;

    public bool Equals(ValueTuple_Default<T1, T2> other) =>
        EqualityComparer<T1>.Default.Equals(Item1, other.Item1) &&
        EqualityComparer<T2>.Default.Equals(Item2, other.Item2);
}
```
With the current mscorlib bits in coreclr, I get repeatable results like this, with each line varying only a little bit:
```
Shared:  0.9470538
Local:   0.6117238
Default: 2.1711197
```
So, the Default case is about 2x the Shared, and 3-4x the Local. However, if I change the mscorlib implementaton of Default to just pre-cache the comparer, rather than using a volatile field and lazy initialization, I get results like this:
```
Shared:  0.9644405
Local:   0.5083141
Default: 0.9415915
```
which suggests that the main difference between using a shared field and ```EqualityComparer<T>.Default``` is the implementation of Default, and in order to get the gains mentioned regarding caching a local comparer (which, per Jan's comments, avoids the generic type hop), we'd actually need to have a comparer field for each T1, T2, etc. on each ValueTuple type, which has the potential to result in a huge number of fields, depending on all the different combinations of types as generic type parameters to the various arities of ```ValueTuple<...>```.

As such, if that's all close to correct, seems to me that if this is a difference we care about, the answer is to change the implementation of ```EqualityComparer<T>.Default```, e.g. https://github.com/dotnet/coreclr/issues/1157, rather than adding such caching here.

jkotas replied:
Agree: dotnet/coreclr#4340

VSadov replied:
Thanks @jkotas 
Considering that change, I think this is a sufficient tradeoff. While additional savings are possible, it does not seem sufficient to justify that many static fields. 
Users sensitive to that might just pass their own comparers to Dictionary.

Will the same changes be applied to the Core FX version?
Like in here:
https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        }
+    }
+
+    public struct ValueTuple<T1> : IEquatable<ValueTuple<T1>>, IStructuralEquatable, IStructuralComparable, IComparable, ITupleInternal
```
As this is implementing IComparable and ```IEquatable<ValueTuple<T1>>```, should it also implement ```IComparable<ValueTuple<T1>>```?

jcouv replied:
My goal was to parallel `System.Tuple` to make it easier to interop. I don't mind implementing the generic IComparable interface too.
I think that's up to corefx team to decide. Let me know.

stephentoub replied:
I think we should implement ```IComparable<T>```, and maybe not implement ```IComparable```.  @weshaggard, @terrajobst, opinions?

weshaggard replied:
I think we should implement both `IComparable` and `IComparable<T>`. @KrzysztofCwalina I couldn't find anything in the framework design guidelines on this particular topic, do you have any thoughts?

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        /// </remarks>
+        public override bool Equals(object other)
+        {
+            return ((IStructuralEquatable)this).Equals(other, EqualityComparer<object>.Default); ;
```
Could this instead be:
```C#
if (!(other is ValueTuple<T1, T2>)) return false;

var otherTuple = (ValueTuple<T1, T2>)other;
return
    EqualityComparer<T1>.Default.Equals(this.Item1, otherTuple.Item1) &&
    EqualityComparer<T2>.Default.Equals(this.Item2, otherTuple.Item2);
```
?

jcouv replied:
In my introduction to these commits, I clarified that only `ValueTuple'1` and `ValueTuple'8` have been fixed with feedback. I want to review those first so I can focus my effort.

The general pattern is that `object.Equals` checks type and delegates to `Equals(ValueTuple)` (which looks as you suggest).

jcouv replied:
Sorry for the confusion. Given that we're converging, my next commit will apply the pattern we agreed to all arities.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        /// </remarks>
+        public bool Equals(ValueTuple<T1, T2> other)
+        {
+            return Equals(Item1, other.Item1) && Equals(Item2, other.Item2);
```
Any reason this can't use ```EqualityComparer<T1>.Default.Equals``` and ```EqualityComparer<T2>.Default.Equals``` rather than ```Object.Equals(Object, Object)```?  The latter is going to box value types, such that if both T1 and T2 were value types, we'll get four unnecessary allocations.

jcouv replied:
`ValueTuple'2` was not fixed yet.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        /// <returns>A 32-bit signed integer hash code.</returns>
+        public override int GetHashCode()
+        {
+            return ((IStructuralEquatable)this).GetHashCode(EqualityComparer<object>.Default);
```
Why are we using the hash code of the equality comparer, rather than a hash code based on Item1 and Item2?  Every ```ValueTuple<T1, T2>``` is going to have the same hash code.

stephentoub replied:
Also, the same comments as earlier about boxing ```this```.  I'll stop commenting on this; just assume that my previous comments apply to all arities :smile:

jcouv replied:
Fixed.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+
+            if (c != 0) return c;
+
+            return comparer.Compare(Item4, objTuple.Item4);
```
Nit: I'd find the logic here a bit easier to follow if the spacing were a bit different:
```C#
int c = comparer.Compare(Item1, objTuple.Item1);
if (c != 0) return c;

c = comparer.Compare(Item2, objTuple.Item2);
if (c != 0) return c;

c = comparer.Compare(Item3, objTuple.Item3);
if (c != 0) return c;

return comparer.Compare(Item4, objTuple.Item4);
```
If you wanted, you could consolidate it even further would something like:
```C#
int c;
if ((c = comparer.Compare(Item1, objTuple.Item1) != 0) return c;
if ((c = comparer.Compare(Item2, objTuple.Item2) != 0) return c;
if ((c = comparer.Compare(Item3, objTuple.Item3) != 0) return c;
return comparer.Compare(Item4, objTuple.Item4);
```

jcouv replied:
I agree. I think I'll go with your first suggestion (grouping pairs of lines).

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        private static readonly EqualityComparer<T7> s_equality7 = EqualityComparer<T7>.Default;
+        private static readonly Comparer<T7> s_comparer7 = Comparer<T7>.Default;
+        private static readonly EqualityComparer<TRest> s_equalityRest = EqualityComparer<TRest>.Default;
+        private static readonly Comparer<TRest> s_comparerRest = Comparer<TRest>.Default;
```
Same question as earlier about static fields... why do we need these?

jcouv replied:
Removed.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        {
+            // We want to have a limited hash in this case.  We'll use the last 8 elements of the tuple
+            ITupleInternal rest = (ITupleInternal)Rest;
+            if (rest.Size >= 8) { return rest.GetHashCode(); }
```
What guarantees that TRest is a tuple?  If it's not, this is going to fail with an InvalidCastException.

jcouv replied:
The `ValueTuple'8` constructor checks that.

```C#
        public ValueTuple(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5, T6 item6, T7 item7, TRest rest)
        {
            if (!(rest is ITupleInternal))
            {
                throw new ArgumentException(SR.ArgumentException_ValueTupleLastArgumentNotAValueTuple);
            }
```

stephentoub replied:
But the ctor won't necessarily be used.  ```ValueTuple`8``` is a struct, so an instance can be used without going through the ctor, and it's mutable, so someone can reasonably write code like:
```C#
var vt = default(ValueTuple<int, int, int, int, int, int, int, int>);
vt.Item8 = 42;
int hc = vt.GetHashCode(); // kaboom
```

jcouv replied:
Ah, yes. I'll fix.

jcouv replied:
On second thought, I think we're fine. `var vt = default(ValueTuple<int, int, int, int, int, int, int, int>);` won't pass the constructor check.

Because all the types that implement `ITupleInternal` are structs, you won't be able to set the `Rest` to something else.

stephentoub replied:
> On second thought, I think we're fine. var vt = default(ValueTuple<int, int, int, int, int, int, int, int>); won't pass the constructor check.

That won't run the ctor.

jcouv replied:
You're right. I must have messed up my test. I'll fix.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+                                                        rest.GetHashCode());
+            }
+
+            Debug.Assert(false, "Missed all cases for computing ValueTuple hash code");
```
What if TRest is ```ValueTuple```?  Then rest.Size will be 0, and k will be 8.

jcouv replied:
Good catch. Fixed.

stephentoub commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+            if (rest.Size >= 8) { return rest.GetHashCode(); }
+
+            // In this case, the rest member has less than 8 elements so we need to combine some our elements with the elements in rest
+            int k = 8 - rest.Size;
```
rest.Size is an interface call.  Consider above calling it once and storing the result into a local so we don't have to call it twice.

jcouv replied:
Done.

stephentoub commented:
In case you have time to improve the test coverage further, you can do a code coverage run locally and see what's missing.  Here's a summary:
![image](https://cloud.githubusercontent.com/assets/2642209/14492627/a555a05a-014e-11e6-8c7f-3e6683ae5a71.png)

jcouv commented:
@mellinoe @stephentoub @tmat @weshaggard @jaredpar 
I pushed a commit to address the second round of feedback. This time I applied the fixes to all the tuples types (sorry for any confusion in last commit).

Do let me know if I should implement the generic `IComparable`.
I have let the version numbers as-is until we're ready to squash and merge.

![image](https://cloud.githubusercontent.com/assets/12466233/14508334/d1153606-017b-11e6-8a25-df97472f9518.png)

KrzysztofCwalina commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+    }
+
+    public struct ValueTuple : IEquatable<ValueTuple>, IStructuralEquatable, IStructuralComparable, IComparable, ITupleInternal
+    {
```
Why does it implement IComparable, but not IComparable<T>?

stephentoub replied:
@KrzysztofCwalina, this is being discussed here: https://github.com/dotnet/corefx/pull/7663/files/77248e5e8e55719dedbb07b9322e9c86106b1492#r59600180

KrzysztofCwalina commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+
+        int ITupleInternal.GetHashCode(IEqualityComparer comparer)
+        {
+            return ((IStructuralEquatable)this).GetHashCode(comparer);
```
This boxes as well.

stephentoub replied:
@KrzysztofCwalina, FYI, I think you're looking at an older commit.  This has since been updated.

jcouv replied:
I'll fix by extracting a common private method instead.

KrzysztofCwalina commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+
+        public override string ToString()
+        {
+            return "()";
```
Is that ToString we want?

jcouv replied:
I think so and it is consistent with how tuples will appear in the IDE. I'm open to suggestions though.

KrzysztofCwalina commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        int Size { get; }
+    }
+
+    public struct ValueTuple : IEquatable<ValueTuple>, IStructuralEquatable, IStructuralComparable, IComparable, ITupleInternal
```
We should have better docs for these types. I think the dcos should explain who these types differ from the old tuples.

jcouv replied:
I have extensive documentation on the `ValueTuple'2` as an example for review. Let me know if that is fine, I will propagate to the types of different arity. 
As for the difference with existing type, I'll add a note on that (again in `ValueTuple'2` for starters).

KrzysztofCwalina commented on src/System.Tuples/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1119 @@
+        {
+            return CombineHashCodes(CombineHashCodes(h1, h2, h3, h4), CombineHashCodes(h5, h6, h7, h8));
+        }
+    }
```
So this type is just a factory for tuples. Why isn't it just a static class?

jcouv replied:
It's also the 0-tuple.

KrzysztofCwalina replied:
Could you point me to an example of how zero-tuple would be used in code? Note that the class tuples don't have zero-tuple.

VSadov replied:
There are no current plans for nuples (0-ple) and womples (1-ple) to be expressible directly in the code, but it keeps coming up in feature discussions that we may have to deal with them at some point. For example there could be a user extendable version of pattern matching based on user defined operators returning tuples. For consistency such scheme may necessitate use of nuples and womples.

Note that there is already a ValueTuple`1, but there is no ValueTuple'0, so this change attempts to add one, if there are no objections, for possible future use.

jcouv commented:
@mellinoe @stephentoub @weshaggard @tmat @jaredpar @KrzysztofCwalina 
The last commit I pushed adds the generic IComparable, removes another instance of boxing.

omariom commented:
Thanks! 
I've been waiting for it [since long ago](http://omari-o.blogspot.ru/2009/09/tuple-as-value-type.html).
With language support is will be just.. what doctor ordered! :tada:

gafter commented:
I see that this implementation lacks the proposed interface `ITuple` to support pattern-matching.

jcouv commented:
@gafter, Yes, I'll add `ITuple` interface and implementation. I got confused.

jcouv commented:
Latest commit adds `ITuple`. 

As discussed with Immo, Krysztof and Wes, we'll want to implement this interface on `System.Tuple`, `KeyValuePair` and any other existing tuple-like type at some point.

KrzysztofCwalina commented:
Could we discuss the shape of the interface as a separate PR? When we met we said the interface is something that we are not yet sure about.

KrzysztofCwalina commented:
I have several concerns about ITuple interface:
1. I don't like that it looks like a collection interface, but it's not.
2. I don't like that it introduces new name for Count/Length.
3. I am not sure this is the best component and/or namespace for this type.
4. I am not thrilled that it boxes. Maybe it's unavoidable, but I would like to discuss. Could you please point me to mathcing/decomposition spec/writeup?

omariom commented:
`ITuple` is implemented by tuples explicitly. It means pattern matching will be boxing them.

gafter commented:
Pattern-matching would only box when the static type of the tuple isn't known to the compiler.

jcouv commented:
Ok, I popped off the commit that contained `ITuple`, as it is not critical for the package I need next week (initial release of tuples to our more publicly visible `future` branch).

Let me know if there is additional feedback or if it's ok to merge (with updated version dependencies and squashing of course).

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+
+            var objTuple = (ValueTuple<T1>)other;
+
+            return Comparer<T1>.Default.Compare(Item1, objTuple.Item1);
```
Nit: in other places I think you've delegated to the generic CompareTo at this point, e.g.
```C#
return CompareTo((ValueTuple<T1>)other);
```

jcouv replied:
Yes, that was on purpose. My thinking for `ValueTuple'1` was to leave the code inline since it is a method call either way.

stephentoub replied:
Ok.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+        /// The <paramref name="other"/> parameter is considered to be equal to the current instance under the following conditions:
+        /// <list type="bullet">
+        ///     <item><description>It is a <seealso cref="ValueTuple{T1, T2}"/> struct.</description></item>
+        ///     <item><description>Its two components are of the same types as the current instance.</description></item>
```
Nit: "are of the same types as the current instance" => "are of the same types as those of the current instance"
(I'm assuming this same thing occurs elsewhere; I won't comment on it there)

jcouv replied:
Fixed here and elsewhere.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+        /// <list type="bullet">
+        ///     <item><description>It is a <seealso cref="ValueTuple{T1, T2}"/> struct.</description></item>
+        ///     <item><description>Its two components are of the same types as the current instance.</description></item>
+        ///     <item><description>Its two components are equal to those of the current instance.Equality is determined by the default object equality comparer for each component.</description></item>
```
Nit: "instance.Equality" => "instance. Equality"

jcouv replied:
Fixed here and in another instance.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+        public override int GetHashCode()
+        {
+            return ValueTuple.CombineHashCodes(EqualityComparer<T1>.Default.GetHashCode(Item1),
+                                                EqualityComparer<T2>.Default.GetHashCode(Item2),
```
Nit: I think in other places you've lined these up, whereas here they're off by one space.

jcouv replied:
Any alignment was incidental, as I only used tabs. New commit will have things neat.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+                && EqualityComparer<T3>.Default.Equals(Item3, objTuple.Item3)
+                && EqualityComparer<T4>.Default.Equals(Item4, objTuple.Item4)
+                && EqualityComparer<T5>.Default.Equals(Item5, objTuple.Item5)
+                && EqualityComparer<T6>.Default.Equals(Item6, objTuple.Item6);
```
Doesn't this need to be using ```comparer``` rather than ```EqualityComparer<T>.Default```?

jcouv replied:
Fixed in every `IStructuralEquatable.Equals` instance.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+                && EqualityComparer<T4>.Default.Equals(Item4, objTuple.Item4)
+                && EqualityComparer<T5>.Default.Equals(Item5, objTuple.Item5)
+                && EqualityComparer<T6>.Default.Equals(Item6, objTuple.Item6)
+                && EqualityComparer<T7>.Default.Equals(Item7, objTuple.Item7);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+
+            var objTuple = (ValueTuple<T1>)other;
+
+            return EqualityComparer<T1>.Default.Equals(Item1, objTuple.Item1);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+            var objTuple = (ValueTuple<T1, T2>)other;
+
+            return EqualityComparer<T1>.Default.Equals(Item1, objTuple.Item1)
+                && EqualityComparer<T2>.Default.Equals(Item2, objTuple.Item2);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

jcouv replied:
Yes. Sorry I missed that. I think it will occur throughout all `IStructuralEquatable.Equals` :-(

stephentoub replied:
Ok, thanks.  Seems like then even with the high code coverage (thanks for that, btw) we're still missing some tests.

jcouv replied:
Fixed.

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+
+            return EqualityComparer<T1>.Default.Equals(Item1, objTuple.Item1)
+                && EqualityComparer<T2>.Default.Equals(Item2, objTuple.Item2)
+                && EqualityComparer<T3>.Default.Equals(Item3, objTuple.Item3);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+            return EqualityComparer<T1>.Default.Equals(Item1, objTuple.Item1)
+                && EqualityComparer<T2>.Default.Equals(Item2, objTuple.Item2)
+                && EqualityComparer<T3>.Default.Equals(Item3, objTuple.Item3)
+                && EqualityComparer<T4>.Default.Equals(Item4, objTuple.Item4);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+                && EqualityComparer<T2>.Default.Equals(Item2, objTuple.Item2)
+                && EqualityComparer<T3>.Default.Equals(Item3, objTuple.Item3)
+                && EqualityComparer<T4>.Default.Equals(Item4, objTuple.Item4)
+                && EqualityComparer<T5>.Default.Equals(Item5, objTuple.Item5);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

stephentoub commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1380 @@
+                && EqualityComparer<T5>.Default.Equals(Item5, objTuple.Item5)
+                && EqualityComparer<T6>.Default.Equals(Item6, objTuple.Item6)
+                && EqualityComparer<T7>.Default.Equals(Item7, objTuple.Item7)
+                && EqualityComparer<TRest>.Default.Equals(Rest, objTuple.Rest);
```
```comparer``` instead of ```EqualityComparer<T>.Default```?

stephentoub commented:
A few more comments, but otherwise LGTM. Thanks!

jcouv commented:
New commit addresses the bug with `IStructuralEquatable.Equals` in all the tuple arities. 
Also addresses comments and alignment feedback.

I added a few test checks and also renamed some test locals to make them more meaningful, but it seems github diff is not very smart about that (sorry).

jcouv commented:
@stephentoub I'm confused by the CI failure. My squashed commit has today's version number. Looking at the test logs, I can't seem to find what is causing the problem.

stephentoub commented:
The tests project.json doesn't look like other tests project.json files I see in the repo, e.g. it's missing some of the sections that are in https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Extensions/tests/project.json

@weshaggard can probably elaborate on why they're necessary, but when I add them to your PR locally, all tests pass.

jcouv commented:
Thanks. It looks like this was changed in a [test runtime overhaul](https://github.com/dotnet/corefx/commit/622ee8f512497fc07621de15e41c6bd575f387c9) after I had last pulled.

jcouv commented on src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs:
```diff
@@ -0,0 +1,1379 @@
+                && comparer.Equals(Item5, objTuple.Item5)
+                && comparer.Equals(Item6, objTuple.Item6)
+                && comparer.Equals(Item7, objTuple.Item7)
+                && comparer.Equals(Rest, objTuple.Rest);
```
@stephentoub I wonder about something. For structural equality and comparison, shouldn't I also use structural equality and comparison on the `Rest`?

The last line here would say: `&& ((IStructuralEquatable)Rest).Equals(objTuple.Rest, comparer);` and I would add some checks before such a cast.
Same question for `IStructuralComparable.CompareTo` below.

stephentoub replied:
What does Tuple do here? Seems we should match it.

jcouv replied:
[System.Tuple](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Tuple.cs#L846) does the same as I implemented. 
But I'm not sure if it's still the right thing to do given that `ValueTuple` types are structs.

stephentoub replied:
How does the class vs struct impact things here?  I'm not understanding why it's a different case, but I could just be missing it.

The StructuralComparers appear to already do this check, so I'd err on the side of just matching what Tuple does:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Collections/StructuralComparisons.cs#L43-L47

jcouv commented:
@stephentoub I'm still getting CI failures without clear diagnostics. I verified that my `project.json` files match the latest ones from `System.Buffers`.

stephentoub commented:
> 

The log shows this:
```
15:37:55 error: file not found: System.ValueTuple.Tests.dll
15:37:55 error: One or more tests failed while running tests from ''.  Exit code 1.
```
I expect it's because the assembly is named incorrectly here:
https://github.com/dotnet/corefx/pull/7663/files#diff-64eb170f29cbbe0ad0c20d343b634c2cR9

stephentoub commented:
Replaced by https://github.com/dotnet/corefx/pull/7803

jcouv commented:
Thanks a bunch.

@weshaggard @stephentoub Let me know the next steps for a signed package to be produced.

FYI, two open issues I'll follow-up on : `ITuple` interface and the problem of weak hashing function.

weshaggard commented:
@jcouv I expect once our official build completes today the package should get published to our http://dotnet.myget.org/F/dotnet-core feed.
